src/librustdoc/clean/types.rs

```diff
@@ -1299,13 +1305,31 @@ crate enum TypeKind {
 }
 
 crate trait GetDefId {
+    /// Use this method to get the [`DefId`] of a [`clean`] AST node.
+    /// This will return [`None`] when called on a primitive [`clean::Type`].
+    /// Use [`Self::def_id_full`] if you are calling it on a primitive [`clean::Type`].
+    ///
+    /// [`clean`]: crate::clean
+    /// [`clean::Type`]: Type
     fn def_id(&self) -> Option<DefId>;
+
+    /// Use this method to get the [`DefId`] of a [`clean`] AST node that may be
+    /// a primitive [`clean::Type`].
+    ///
+    /// See [`Self::def_id`] for more.
```

Comment:
I think we should say ``including primitive [`clean::Type`]s`` because otherwise people may think we're talking about calling this on e.g. `u32`, rather than e.g. `clean::Type::Primitive(PrimitiveType::U32)`.

Reply:
But `u32` is a `PrimitiveType`. I think because this is in `clean/types.rs`, no one will be confused that this is referring to `hir::Ty` or `ty::Ty` (and note that even that's ambiguous, rustdoc isn't special here).

Reply:
> But `u32` is a `PrimitiveType`. I think because this is in `clean/types.rs`, no one will be confused that this is referring to `hir::Ty` or `ty::Ty` (and note that even that's ambiguous, rustdoc isn't special here).

That's not what I was referring to me; I worry there's a change that people (e.g. who are new to rustdoc) may initially interpret the docs you're suggesting as saying that you can e.g. call `15_u32.def_id_full()`. Maybe the best would be:

> Use this method to get the [`DefId`] of a [`clean`] AST node, including [`PrimitiveType`]s.

Reply:
I like the last suggestion of @camelid so I switched to it.